### PR TITLE
optimizer/plan: support index hint.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: optimizer/plan/planbuilder.go [hunk 1]
```diff
@@ -324,16 +324,84 @@ func (b *planBuilder) buildSingleTable(sel *ast.SelectStmt) Plan {
...
+		if hint.HintScope == ast.HintForScan {
+			usableHints = append(usableHints, hint)
+		}
+	}
+	if len(usableHints) == 0 {
+		return table.TableInfo.Indices, true
+	}
+	var hasUse bool
+	for _, hint := range usableHints {
+		// Currently we don't distinguish Force and Use because our cost estimation is not reliable.
```
Comment (shenli) [inline optimizer/plan/planbuilder.go]: "distinguish between Force and Use" or "distinguish Force from Use"

File: optimizer/plan/planbuilder.go [hunk 2]
```diff
@@ -324,16 +324,84 @@ func (b *planBuilder) buildSingleTable(sel *ast.SelectStmt) Plan {
...
+					indices = append(indices, idx)
+				}
+			}
+		}
+	}
+	// If we have got FORCE or USE index hint, table scan is excluded.
+	if len(indices) != 0 {
+		return indices, false
+	}
+	if hasUse {
```
Comment (shenli) [inline optimizer/plan/planbuilder.go]: If usableHints is empty, we will not use any index?

File: optimizer/plan/planbuilder.go [hunk 3]
```diff
@@ -324,16 +324,84 @@ func (b *planBuilder) buildSingleTable(sel *ast.SelectStmt) Plan {
...
+		// Empty use hint means don't use any index.
+		return nil, true
+	}
+
+	var ignores []*model.IndexInfo
+	// Collect all the ignore index hints.
+	for _, hint := range usableHints {
```
Comment (shenli) [inline optimizer/plan/planbuilder.go]: Can we merge this loop with the above one?

File: optimizer/plan/planbuilder.go (hunk 2 above)
Comment (coocood) [inline optimizer/plan/planbuilder.go]: It returns in line 348.

**Comment by coocood** [comment]
@shenli PTAL

**Comment by ngaut** [comment]
PTAL @hanfei19910905 @zimulala

**Comment by zimulala** [comment]
Add a comment about that `IGNORE INDEX` is applied over the result of the previous step. For example, the following two queries are equivalent:
```
SELECT * FROM t1 USE INDEX (i1) IGNORE INDEX (i2) USE INDEX (i2);
SELECT * FROM t1 USE INDEX (i1);
```

**Comment by zimulala** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**